// @sanity/base/lib/components/previews/inlinePreview.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.InlinePreview = void 0;

var _react = _interopRequireWildcard(require("react"));

var _useDevicePixelRatio = require("use-device-pixel-ratio");

var _inlinePreview = require("./inlinePreview.styled");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var DEFAULT_MEDIA_DIMENSIONS = {
  width: 32,
  height: 32,
  fit: 'crop',
  aspect: 1,
  dpr: (0, _useDevicePixelRatio.getDevicePixelRatio)()
};

var InlinePreview = props => {
  var title = props.title,
      _props$fallbackTitle = props.fallbackTitle,
      fallbackTitle = _props$fallbackTitle === void 0 ? 'Untitled' : _props$fallbackTitle,
      media = props.media,
      _props$mediaDimension = props.mediaDimensions,
      mediaDimensions = _props$mediaDimension === void 0 ? DEFAULT_MEDIA_DIMENSIONS : _props$mediaDimension;

  var _media = (0, _react.useMemo)(() => {
    if (typeof media === 'function') {
      return media({
        dimensions: mediaDimensions,
        layout: 'inline'
      });
    }

    return media;
  }, [media, mediaDimensions]);

  return /*#__PURE__*/_react.default.createElement(_inlinePreview.RootSpan, {
    "data-testid": "inline-preview"
  }, _media && /*#__PURE__*/_react.default.createElement(_inlinePreview.MediaSpan, {
    "data-testid": "inline-preview-media"
  }, _media, /*#__PURE__*/_react.default.createElement("span", null)), /*#__PURE__*/_react.default.createElement(_inlinePreview.TextSpan, {
    "data-testid": "inline-preview-title",
    size: 1
  }, title || fallbackTitle));
};

exports.InlinePreview = InlinePreview;
