// @sanity/desk-tool/lib/components/paneItem/PaneItemPreview.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PaneItemPreview = void 0;

var _preview = require("part:@sanity/base/preview");

var _react = _interopRequireDefault(require("react"));

var _helpers = require("./helpers");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class PaneItemPreview extends _react.default.Component {
  constructor(props) {
    super(props);

    _defineProperty(this, "state", {});

    _defineProperty(this, "subscription", void 0);

    var value = props.value,
        schemaType = props.schemaType;
    var title = value.title;
    var sync = true;
    this.subscription = (0, _helpers.getPreviewStateObservable)(schemaType, value._id, title).subscribe(state => {
      if (sync) {
        this.state = state;
      } else {
        this.setState(state);
      }
    });
    sync = false;
  }

  componentWillUnmount() {
    if (this.subscription) {
      this.subscription.unsubscribe();
    }
  }

  render() {
    var _this$props = this.props,
        value = _this$props.value,
        layout = _this$props.layout,
        icon = _this$props.icon;
    var _this$state = this.state,
        draft = _this$state.draft,
        published = _this$state.published,
        isLoading = _this$state.isLoading;
    return /*#__PURE__*/_react.default.createElement(_preview.SanityDefaultPreview, {
      value: (0, _helpers.getValueWithFallback)({
        value,
        draft,
        published
      }),
      isPlaceholder: isLoading,
      icon: icon,
      layout: layout,
      status: isLoading ? null : (0, _helpers.getStatusIndicator)(draft, published)
    });
  }

}

exports.PaneItemPreview = PaneItemPreview;
