// @sanity/form-builder/lib/inputs/arrays/ArrayOfObjectsInput/item/RowItem.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RowItem = void 0;

var _presence = require("@sanity/base/presence");

var _react = _interopRequireDefault(require("react"));

var _ui = require("@sanity/ui");

var _icons = require("@sanity/icons");

var _components = require("@sanity/base/components");

var _autoId = require("@reach/auto-id");

var _Preview = _interopRequireDefault(require("../../../../Preview"));

var _DragHandle = require("../../common/DragHandle");

var _ItemWithMissingType = require("./ItemWithMissingType");

var _RowWrapper = require("./components/RowWrapper");

var _excluded = ["isSortable", "value", "onClick", "onKeyPress", "onFocus", "type", "readOnly", "presence", "onRemove", "validation"];

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var dragHandle = /*#__PURE__*/_react.default.createElement(_DragHandle.DragHandle, {
  paddingX: 1,
  paddingY: 3
});

var RowItem = /*#__PURE__*/_react.default.forwardRef(function RegularItem(props, ref) {
  var _type$options;

  var focusRef = _react.default.useRef(null);

  var isSortable = props.isSortable,
      value = props.value,
      onClick = props.onClick,
      onKeyPress = props.onKeyPress,
      onFocus = props.onFocus,
      type = props.type,
      readOnly = props.readOnly,
      presence = props.presence,
      onRemove = props.onRemove,
      validation = props.validation,
      rest = _objectWithoutProperties(props, _excluded);

  var hasErrors = validation.some(v => v.level === 'error');
  var hasWarnings = validation.some(v => v.level === 'warning');
  var id = (0, _autoId.useId)();
  return /*#__PURE__*/_react.default.createElement(_RowWrapper.RowWrapper, _extends({}, rest, {
    ref: ref,
    radius: 2,
    padding: 1,
    tone: readOnly ? 'transparent' : hasErrors ? 'critical' : hasWarnings ? 'caution' : 'default'
  }), /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    align: "center"
  }, isSortable && /*#__PURE__*/_react.default.createElement(_ui.Card, {
    className: "dragHandle",
    tone: "inherit",
    marginRight: 1
  }, dragHandle), type ? /*#__PURE__*/_react.default.createElement(_ui.Card, {
    as: "button",
    type: "button",
    tone: "inherit",
    radius: 2,
    padding: 1,
    flex: 1,
    onClick: onClick,
    ref: focusRef,
    onKeyPress: onKeyPress,
    onFocus: onFocus,
    __unstable_focusRing: true
  }, /*#__PURE__*/_react.default.createElement(_Preview.default, {
    layout: ((_type$options = type.options) === null || _type$options === void 0 ? void 0 : _type$options.layout) === 'grid' ? 'media' : 'default',
    value: value,
    type: type
  })) : /*#__PURE__*/_react.default.createElement(_ui.Box, {
    flex: 1
  }, /*#__PURE__*/_react.default.createElement(_ItemWithMissingType.ItemWithMissingType, {
    value: value,
    onFocus: onFocus
  })), /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    align: "center"
  }, !readOnly && presence.length > 0 && /*#__PURE__*/_react.default.createElement(_ui.Box, {
    marginLeft: 1
  }, /*#__PURE__*/_react.default.createElement(_presence.FieldPresence, {
    presence: presence,
    maxAvatars: 1
  })), validation.length > 0 && /*#__PURE__*/_react.default.createElement(_ui.Box, {
    marginLeft: 1,
    paddingX: 1,
    paddingY: 3
  }, /*#__PURE__*/_react.default.createElement(_components.FormFieldValidationStatus, {
    __unstable_markers: validation,
    __unstable_showSummary: !value._ref
  })), /*#__PURE__*/_react.default.createElement(_ui.MenuButton, {
    button: /*#__PURE__*/_react.default.createElement(_ui.Button, {
      padding: 2,
      mode: "bleed",
      icon: _icons.EllipsisVerticalIcon
    }),
    id: "".concat(id, "-menuButton"),
    menu: /*#__PURE__*/_react.default.createElement(_ui.Menu, null, !readOnly && /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_ui.MenuItem, {
      text: "Remove",
      tone: "critical",
      icon: _icons.TrashIcon,
      onClick: onRemove
    }))),
    placement: "right",
    popover: {
      portal: true,
      tone: 'default'
    }
  }), !value._key && /*#__PURE__*/_react.default.createElement(_ui.Box, {
    marginLeft: 1
  }, /*#__PURE__*/_react.default.createElement(_ui.Tooltip, {
    content: /*#__PURE__*/_react.default.createElement(_ui.Box, {
      padding: 2
    }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
      muted: true,
      size: 1
    }, "This item is missing the required ", /*#__PURE__*/_react.default.createElement("code", null, "_key"), " property.")),
    placement: "top"
  }, /*#__PURE__*/_react.default.createElement(_ui.Badge, {
    mode: "outline",
    tone: "caution"
  }, "Missing key"))))));
});

exports.RowItem = RowItem;
