// @sanity/base/lib/components/previews/__workshop__/BlockImagePreviewStory.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = BlockImagePreviewStory;

var _icons = require("@sanity/icons");

var _ui = require("@sanity/ui");

var _uiWorkshop = require("@sanity/ui-workshop");

var _react = _interopRequireDefault(require("react"));

var _blockImagePreview = require("../blockImagePreview");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function BlockImagePreviewStory() {
  var title = (0, _uiWorkshop.useString)('Title', 'Title', 'Props');
  var subtitle = (0, _uiWorkshop.useString)('Subtitle', 'Subtitle', 'Props');
  var description = (0, _uiWorkshop.useString)('Description', undefined, 'Props');
  var withImage = (0, _uiWorkshop.useBoolean)('With image', true, 'Props');
  var withStatus = (0, _uiWorkshop.useBoolean)('With status', false, 'Props');
  var withActions = (0, _uiWorkshop.useBoolean)('With actions', true, 'Props');
  return /*#__PURE__*/_react.default.createElement(_ui.Card, {
    height: "fill"
  }, /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    align: "center",
    height: "fill",
    justify: "center",
    padding: 4,
    sizing: "border"
  }, /*#__PURE__*/_react.default.createElement(_ui.Container, {
    width: 1
  }, /*#__PURE__*/_react.default.createElement(_ui.Card, {
    border: true
  }, /*#__PURE__*/_react.default.createElement(_blockImagePreview.BlockImagePreview, {
    description: description,
    actions: withActions && /*#__PURE__*/_react.default.createElement(_ui.MenuButton, {
      id: "id",
      button: /*#__PURE__*/_react.default.createElement(_ui.Button, {
        fontSize: 1,
        icon: _icons.EllipsisVerticalIcon,
        mode: "bleed"
      }),
      menu: /*#__PURE__*/_react.default.createElement(_ui.Menu, null, /*#__PURE__*/_react.default.createElement(_ui.MenuItem, {
        text: "Edit",
        icon: _icons.EditIcon
      }), /*#__PURE__*/_react.default.createElement(_ui.MenuItem, {
        text: "Delete",
        icon: _icons.TrashIcon,
        tone: "critical"
      })),
      popover: {
        constrainSize: true
      }
    }),
    media: withImage && /*#__PURE__*/_react.default.createElement("img", {
      src: "https://source.unsplash.com/600x600"
    }),
    status: withStatus && /*#__PURE__*/_react.default.createElement(_ui.Text, {
      muted: true,
      size: 1
    }, /*#__PURE__*/_react.default.createElement(_icons.EditIcon, null)),
    subtitle: subtitle,
    title: title
  })))));
}
