// @sanity/base/lib/components/previews/__workshop__/DefaultPreviewStory.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = DefaultPreviewStory;

var _icons = require("@sanity/icons");

var _ui = require("@sanity/ui");

var _uiWorkshop = require("@sanity/ui-workshop");

var _react = _interopRequireDefault(require("react"));

var _defaultPreview = require("../defaultPreview");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function DefaultPreviewStory() {
  var title = (0, _uiWorkshop.useString)('Title', 'Title', 'Props');
  var subtitle = (0, _uiWorkshop.useString)('Subtitle', 'Subtitle', 'Props');
  var selected = (0, _uiWorkshop.useBoolean)('Selected', false, 'Props');
  var withImage = (0, _uiWorkshop.useBoolean)('With image', false, 'Props');
  return /*#__PURE__*/_react.default.createElement(_ui.Card, {
    height: "fill"
  }, /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    align: "center",
    height: "fill",
    justify: "center",
    padding: 4,
    sizing: "border"
  }, /*#__PURE__*/_react.default.createElement(_ui.Container, {
    width: 0
  }, /*#__PURE__*/_react.default.createElement(_ui.Card, {
    "data-as": "button",
    padding: 2,
    radius: 2,
    selected: selected
  }, /*#__PURE__*/_react.default.createElement(_defaultPreview.DefaultPreview, {
    media: withImage ? /*#__PURE__*/_react.default.createElement("img", {
      src: "https://source.unsplash.com/35x35/?abstract"
    }) : /*#__PURE__*/_react.default.createElement(_icons.DocumentIcon, null),
    status: /*#__PURE__*/_react.default.createElement(_ui.Text, {
      muted: true
    }, /*#__PURE__*/_react.default.createElement(_icons.EditIcon, null)),
    subtitle: subtitle,
    title: title
  })))));
}
