// @sanity/default-layout/lib/navbar/loginStatus/LoginStatus.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LoginStatus = LoginStatus;

var _react = _interopRequireDefault(require("react"));

var _components = require("@sanity/base/components");

var _ui = require("@sanity/ui");

var _icons = require("@sanity/icons");

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _templateObject;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

var StyledMenu = (0, _styledComponents.default)(_ui.Menu)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  min-width: 125px;\n"])));
var MENU_BUTTON_POPOVER_PROPS = {
  portal: true,
  scheme: 'light',
  placement: 'bottom-end'
};

function LoginStatus(_ref) {
  var onLogout = _ref.onLogout;
  return /*#__PURE__*/_react.default.createElement(_ui.MenuButton, {
    button: /*#__PURE__*/_react.default.createElement(_ui.Button, {
      mode: "bleed",
      paddingX: 1,
      paddingY: 0,
      title: "Toggle user menu"
    }, /*#__PURE__*/_react.default.createElement(_ui.Flex, {
      align: "center"
    }, /*#__PURE__*/_react.default.createElement(_components.UserAvatar, {
      size: "medium",
      tone: "navbar",
      userId: "me"
    }), /*#__PURE__*/_react.default.createElement(_ui.Box, {
      marginLeft: 1
    }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
      size: 1
    }, /*#__PURE__*/_react.default.createElement(_icons.ChevronDownIcon, null))))),
    id: "login-status-menu",
    menu: /*#__PURE__*/_react.default.createElement(StyledMenu, null, /*#__PURE__*/_react.default.createElement(_ui.MenuItem, {
      text: "Sign out",
      iconRight: _icons.LeaveIcon,
      onClick: onLogout
    })),
    popover: MENU_BUTTON_POPOVER_PROPS
  });
}
