// @sanity/base/lib/components/collapseMenu/CollapseMenu.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CollapseMenu = CollapseMenu;

var _react = _interopRequireWildcard(require("react"));

var _ui = require("@sanity/ui");

var _reactIntersectionObserver = require("react-intersection-observer");

var _icons = require("@sanity/icons");

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _ = require(".");

var _excluded = ["buttonProps", "collapseText", "dividerBefore", "icon", "menuItemProps", "selected", "text", "tooltipProps"],
    _excluded2 = ["buttonProps", "dividerBefore", "menuItemProps", "tooltipProps"],
    _excluded3 = ["buttonProps", "dividerBefore", "collapseText", "menuItemProps", "text", "tooltipProps"];

var _templateObject, _templateObject2, _templateObject3;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

var RootBox = (0, _styledComponents.default)(_ui.Box)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  border-radius: inherit;\n  position: relative;\n  overflow: hidden;\n  /**\n  * We need to add padding + negative margin to make the focusRing  visible\n  */\n  padding: ", "px;\n  margin: -", "px;\n"])), _ref => {
  var theme = _ref.theme;
  return theme.sanity.focusRing.width;
}, _ref2 => {
  var theme = _ref2.theme;
  return theme.sanity.focusRing.width;
});
var InnerFlex = (0, _styledComponents.default)(_ui.Flex)(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["\n  position: absolute;\n  top: 0;\n  left: 0;\n  right: 0;\n  bottom: 0;\n  overflow: hidden;\n  white-space: nowrap;\n  pointer-events: ", ";\n  opacity: ", ";\n  position: ", ";\n  visibility: ", ";\n  width: ", ";\n  /**\n  * We need to add padding + negative margin to make the focusRing  visible\n  */\n  padding: ", "px;\n  margin: -", "px;\n"])), _ref3 => {
  var $hide = _ref3.$hide;
  return $hide ? 'none' : 'inherit';
}, _ref4 => {
  var $hide = _ref4.$hide;
  return $hide ? 0 : 1;
}, _ref5 => {
  var $hide = _ref5.$hide;
  return $hide ? 'absolute' : 'static';
}, _ref6 => {
  var $hide = _ref6.$hide;
  return $hide ? 'hidden' : 'visible';
}, _ref7 => {
  var $hide = _ref7.$hide;
  return $hide ? 'max-content' : 'auto';
}, _ref8 => {
  var theme = _ref8.theme;
  return theme.sanity.focusRing.width;
}, _ref9 => {
  var theme = _ref9.theme;
  return theme.sanity.focusRing.width;
});
var OptionBox = (0, _styledComponents.default)(_ui.Box)(_templateObject3 || (_templateObject3 = _taggedTemplateLiteral(["\n  list-style: none;\n  display: flex;\n  white-space: nowrap;\n  visibility: ", ";\n  pointer-events: ", ";\n"])), _ref10 => {
  var $inView = _ref10.$inView;
  return $inView ? 'visible' : 'hidden';
}, _ref11 => {
  var $inView = _ref11.$inView;
  return $inView ? 'inherit' : 'none';
});

function MenuCollapse(_ref12) {
  var menuButton = _ref12.menuButton,
      popoverProps = _ref12.popoverProps,
      options = _ref12.options,
      disableRestoreFocusOnClose = _ref12.disableRestoreFocusOnClose,
      onClose = _ref12.onClose;
  var handleClose = (0, _react.useCallback)(() => {
    if (onClose) {
      onClose();
    }
  }, [onClose]);
  return /*#__PURE__*/_react.default.createElement(_ui.MenuButton, {
    button: menuButton,
    id: "collapse-menu",
    popover: popoverProps,
    __unstable_disableRestoreFocusOnClose: disableRestoreFocusOnClose,
    onClose: handleClose,
    menu: /*#__PURE__*/_react.default.createElement(_ui.Menu, null, options.map((child, index) => {
      var _ref13 = child.props,
          buttonProps = _ref13.buttonProps,
          collapseText = _ref13.collapseText,
          dividerBefore = _ref13.dividerBefore,
          icon = _ref13.icon,
          _ref13$menuItemProps = _ref13.menuItemProps,
          menuItemProps = _ref13$menuItemProps === void 0 ? {} : _ref13$menuItemProps,
          selected = _ref13.selected,
          text = _ref13.text,
          tooltipProps = _ref13.tooltipProps,
          restProps = _objectWithoutProperties(_ref13, _excluded);

      return /*#__PURE__*/_react.default.createElement(_react.Fragment, {
        key: child.key
      }, dividerBefore && index !== 0 && /*#__PURE__*/_react.default.createElement(_ui.MenuDivider, null), /*#__PURE__*/_react.default.createElement(_ui.MenuItem, _extends({}, restProps, {
        icon: icon,
        text: text,
        fontSize: 2,
        radius: 2
      }, menuItemProps, {
        pressed: selected
      })));
    }))
  });
}

function CollapseMenu(props) {
  var children = props.children,
      shouldCollapse = props.collapsed,
      _props$gap = props.gap,
      gap = _props$gap === void 0 ? 1 : _props$gap,
      menuButton = props.menuButton,
      menuPopoverProps = props.menuPopoverProps,
      onMenuClose = props.onMenuClose,
      onMenuVisible = props.onMenuVisible,
      _props$disableRestore = props.disableRestoreFocusOnClose,
      disableRestoreFocusOnClose = _props$disableRestore === void 0 ? false : _props$disableRestore;

  var _useState = (0, _react.useState)(null),
      _useState2 = _slicedToArray(_useState, 2),
      rootBoxElement = _useState2[0],
      setRootBoxElement = _useState2[1];

  var _useState3 = (0, _react.useState)(null),
      _useState4 = _slicedToArray(_useState3, 2),
      innerFlexElement = _useState4[0],
      setInnerFlexElement = _useState4[1];

  var _useState5 = (0, _react.useState)(true),
      _useState6 = _slicedToArray(_useState5, 2),
      collapsed = _useState6[0],
      setCollapsed = _useState6[1];

  var _useState7 = (0, _react.useState)([]),
      _useState8 = _slicedToArray(_useState7, 2),
      menuOptions = _useState8[0],
      setMenuOptions = _useState8[1];

  var rootBoxRect = (0, _ui.useElementRect)(rootBoxElement);
  /**
   * Array of children
   */

  var childrenArray = (0, _react.useMemo)(() => _react.Children.toArray(children), [children]);
  /**
   * Filter to get the latest state of menu options
   */

  var menuOptionsArray = (0, _react.useMemo)(() => childrenArray.filter(_ref14 => {
    var key = _ref14.key;
    return menuOptions.find(o => o.key === key);
  }), [childrenArray, menuOptions]);
  /**
   * Menu popover props
   */

  var popoverProps = (0, _react.useMemo)(() => _objectSpread({
    portal: true,
    placement: 'bottom',
    preventOverflow: true
  }, menuPopoverProps), [menuPopoverProps]);
  /**
   * Pick what button to render as menu button
   */

  var menuButtonToRender = (0, _react.useMemo)(() => {
    if (menuButton) {
      return menuButton;
    }

    return /*#__PURE__*/_react.default.createElement(_ui.Button, {
      mode: "bleed",
      icon: _icons.EllipsisVerticalIcon
    });
  }, [menuButton]);
  /**
   * Check if menu is visible
   */

  var menuIsVisible = (0, _react.useMemo)(() => collapsed && menuOptionsArray.length > 0, [collapsed, menuOptionsArray.length]);
  var handleMenuClose = (0, _react.useCallback)(() => {
    if (onMenuClose) {
      onMenuClose();
    }
  }, [onMenuClose]);
  (0, _react.useEffect)(() => {
    if (onMenuVisible) {
      onMenuVisible(menuIsVisible);
    }
  }, [menuIsVisible, onMenuVisible]);
  /**
   * Compare RootBox width with InnerFlex scrollWidth to collapse/expand menu
   */

  (0, _react.useEffect)(() => {
    if (rootBoxRect && innerFlexElement) {
      var collapse = rootBoxRect.width < innerFlexElement.scrollWidth;
      setCollapsed(collapse);
    }
  }, [innerFlexElement, rootBoxRect]); // Add or remove option in menuOptions

  var handleInViewChange = (0, _react.useCallback)(payload => {
    var child = payload.child,
        inView = payload.inView;
    var exists = menuOptions.some(_ref15 => {
      var key = _ref15.key;
      return key === child.key;
    });

    if (!inView && !exists) {
      setMenuOptions(prev => [child, ...prev]);
    }

    if (inView && exists) {
      var updatedOptions = menuOptions.filter(_ref16 => {
        var key = _ref16.key;
        return key !== child.key;
      });
      setMenuOptions(updatedOptions);
    }
  }, [menuOptions]); //Check if child is in menu

  var isInMenu = (0, _react.useCallback)(childKey => {
    var exists = menuOptionsArray.some(o => o.key === childKey);
    return exists;
  }, [menuOptionsArray]);
  /**
   * Return the collapsed menu
   */

  if (shouldCollapse) {
    return /*#__PURE__*/_react.default.createElement(MenuCollapse, {
      options: childrenArray,
      menuButton: menuButtonToRender,
      popoverProps: popoverProps,
      disableRestoreFocusOnClose: disableRestoreFocusOnClose,
      onClose: handleMenuClose
    });
  }

  return /*#__PURE__*/_react.default.createElement(RootBox, {
    ref: setRootBoxElement,
    display: "flex",
    "data-ui": "CollapseMenu"
  }, /*#__PURE__*/_react.default.createElement(InnerFlex, {
    align: "center",
    ref: setInnerFlexElement,
    $hide: collapsed,
    "aria-hidden": collapsed,
    "data-testid": "inner-flex-expanded"
  }, /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    as: "ul",
    gap: gap
  }, childrenArray.map((child, index) => {
    var _ref17 = child.props,
        _ref17$buttonProps = _ref17.buttonProps,
        buttonProps = _ref17$buttonProps === void 0 ? {} : _ref17$buttonProps,
        dividerBefore = _ref17.dividerBefore,
        menuItemProps = _ref17.menuItemProps,
        tooltipProps = _ref17.tooltipProps,
        restProps = _objectWithoutProperties(_ref17, _excluded2);

    var showDivider = dividerBefore && index !== 0;
    return /*#__PURE__*/_react.default.createElement(_react.Fragment, {
      key: child.key
    }, showDivider && /*#__PURE__*/_react.default.createElement(_.CollapseMenuDivider, null), /*#__PURE__*/_react.default.createElement(_ui.Tooltip, _extends({
      portal: true
    }, tooltipProps, {
      disabled: (tooltipProps === null || tooltipProps === void 0 ? void 0 : tooltipProps.disabled) || !(tooltipProps !== null && tooltipProps !== void 0 && tooltipProps.text),
      content: /*#__PURE__*/_react.default.createElement(_ui.Box, {
        padding: 2,
        sizing: "border"
      }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
        size: 1,
        muted: true
      }, tooltipProps === null || tooltipProps === void 0 ? void 0 : tooltipProps.text))
    }), /*#__PURE__*/_react.default.createElement(_ui.Box, {
      as: "li"
    }, /*#__PURE__*/(0, _react.cloneElement)(child, _objectSpread({}, _objectSpread(_objectSpread({}, restProps), buttonProps))))));
  }))), /*#__PURE__*/_react.default.createElement(InnerFlex, {
    align: "center",
    gap: gap,
    $hide: !collapsed,
    "aria-hidden": !collapsed,
    "data-testid": "inner-flex-collapsed"
  }, /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    gap: gap,
    as: "ul"
  }, childrenArray.map((child, index) => {
    if (child.type !== _.CollapseMenuButton) {
      return null;
    }

    var _ref18 = child.props,
        _ref18$buttonProps = _ref18.buttonProps,
        buttonProps = _ref18$buttonProps === void 0 ? {} : _ref18$buttonProps,
        dividerBefore = _ref18.dividerBefore,
        _ref18$collapseText = _ref18.collapseText,
        collapseText = _ref18$collapseText === void 0 ? true : _ref18$collapseText,
        menuItemProps = _ref18.menuItemProps,
        text = _ref18.text,
        tooltipProps = _ref18.tooltipProps,
        restProps = _objectWithoutProperties(_ref18, _excluded3);

    var showDivider = index !== 0 && dividerBefore;
    return /*#__PURE__*/_react.default.createElement(_reactIntersectionObserver.InView, {
      "aria-hidden": isInMenu(child.key) // eslint-disable-next-line react/jsx-no-bind
      ,
      onChange: inView => handleInViewChange({
        inView,
        child
      }),
      key: child.key,
      threshold: 1,
      rootMargin: "1px",
      root: rootBoxElement
    }, _ref19 => {
      var ref = _ref19.ref,
          inView = _ref19.inView;
      return /*#__PURE__*/_react.default.createElement(OptionBox, {
        ref: ref,
        as: "li",
        $inView: inView && collapsed
      }, showDivider && /*#__PURE__*/_react.default.createElement(_.CollapseMenuDivider, null), /*#__PURE__*/_react.default.createElement(_ui.Box, {
        marginLeft: showDivider ? gap : undefined
      }, /*#__PURE__*/_react.default.createElement(_ui.Tooltip, _extends({
        disabled: !inView || !collapseText,
        portal: true
      }, tooltipProps, {
        content: /*#__PURE__*/_react.default.createElement(_ui.Box, {
          padding: 2,
          sizing: "border"
        }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
          size: 1
        }, (tooltipProps === null || tooltipProps === void 0 ? void 0 : tooltipProps.text) || text))
      }), /*#__PURE__*/(0, _react.cloneElement)(child, _objectSpread(_objectSpread({}, _objectSpread(_objectSpread({}, restProps), buttonProps)), {}, {
        'aria-label': text,
        text: collapseText ? null : text
      })))));
    });
  }))), menuIsVisible && /*#__PURE__*/_react.default.createElement(MenuCollapse, {
    options: menuOptionsArray,
    menuButton: menuButtonToRender,
    popoverProps: popoverProps,
    disableRestoreFocusOnClose: disableRestoreFocusOnClose,
    onClose: handleMenuClose
  }));
}
